rsyslog bugfix: fixes rsyslog segfault during shutdown when relp is configured with TLS

sourav-sharma796 opened the pull request:
Hi,

I've been working on an issue where `rsyslog` generates a coredump. 

The problem occurs when `rsyslog` is using the `relp` module with TLS configured.

The issue arises when `rsyslog` is shutting down. As part of the shutdown, `rsyslog` starts unloading and destroying modules. Since we've configured the `relp` module, `rsyslog` calls `relpEngineDestruct()` from the `modExit()` function. This triggers the destruction of TCP sessions in the `relpTcpDestruct()` function, which then calls `relpTcpDestructTLS()` to handle the shutdown of SSL connections. In this case, the `relpTcpDestructTLS_ossl()` function is used to shut down SSL connections with OpenSSL.

During the shutdown process, if there is an issue with the connection, the SSL shutdown fails, causing an `SSL_ERROR_SSL` error to be logged by the `onErr()` function. By this point, the `freeCnf()` function has already freed variables like `pszBindPort`, which are needed for logging the error. This leads to a segfault when the `pszBindPort` variable is accessed, triggering a  coredump.

To resolve the issue, I added a check to prevent the segmentation fault. I introduced a global variable initialized to zero, which is set to one if the configuration has been freed. In the `onErr()` function, when an error is logged, the `pszBindPort` variable is omitted if the global variable indicates that the configuration has already been freed.

Thanks!

<!--
LEGAL GDPR NOTICE:
According to the European data protection laws (GDPR), we would like to make you
aware that contributing to rsyslog via git will permanently store the
name and email address you provide as well as the actual commit and the
time and date you made it inside git's version history. This is inevitable,
because it is a main feature git. If you are concerned about your
privacy, we strongly recommend to use

--author "anonymous <gdpr@example.com>"

together with your commit. Also please do NOT sign your commit in this case,
as that potentially could lead back to you. Please note that if you use your
real identity, the GDPR grants you the right to have this information removed
later. However, we have valid reasons why we cannot remove that information
later on. The reasons are:

* this would break git history and make future merges unworkable
* the rsyslog projects has legitimate interest to keep a permanent record of the
  contributor identity, once given, for
  - copyright verification
  - being able to provide proof should a malicious commit be made

Please also note that your commit is public and as such will potentially be
processed by many third-parties. Git's distributed nature makes it impossible
to track where exactly your commit, and thus your personal data, will be stored
and be processed. If you would not like to accept this risk, please do either
commit anonymously or refrain from contributing to the rsyslog project.
-->

sourav-sharma796 commented:
Hi,
@rgerhards , can you please review this PR?

Thanks!

sourav-sharma796 commented:
Hi,
@rgerhards , can you please review this PR?

Thanks!

rgerhards commented:
/gemini review

rgerhards commented:
sry for the unresponsiveness, now I am back :-)

sourav-sharma796 commented on plugins/imrelp/imrelp.c:
```diff
@@ -827,6 +836,7 @@ CODESTARTfreeCnf
 		inst = inst->next;
 		free(del);
 	}
+	confFreed=1;
```
Yes, I will move it in the beginning

sourav-sharma796 replied:
Fixed in https://github.com/rsyslog/rsyslog/pull/5600/commits/732db05ff6d396ae3f9692d4f766cf90c1a4b51c

sourav-sharma796 commented:
Hi, @rgerhards can you please take a look?

sourav-sharma796 commented:
Hi, @rgerhards can you please take a look?

sourav-sharma796 commented:
Hi @rgerhards,

can you please check this?

Thanks!

sourav-sharma796 commented:
Hi @rgerhards,

can you please check this?

Thanks!

rgerhards commented:
I had problems with notifications, so I only noticed now. It would be great if you could rebase so that we can see CI results.

rgerhards commented:
disregard my comemnt. Looks like I had not all page updates.

rgerhards commented:
/gemini review

sourav-sharma796 commented on plugins/imrelp/imrelp.c:
```diff
@@ -59,6 +59,8 @@ MODULE_CNFNAME("imrelp")
 DEF_IMOD_STATIC_DATA;
 DEFobjCurrIf(net) DEFobjCurrIf(prop) DEFobjCurrIf(ruleset) DEFobjCurrIf(glbl) DEFobjCurrIf(statsobj)
 
+int confFreed=0;
```
Will correct this asap

sourav-sharma796 replied:
Hi,
As per your suggestion, I tried adding confFreed variable inside modConfData_s struct. But that will require calloc of this struct pointer which will be like wasting memory. I tried using loadModConf which gets memory from pModConf in BEGINbeginCnfLoad but BEGINendCnfLoad is getting called after some time due to which when onErr is called loadModConf becomes NULL.

I think confFreed can be used like bLegacyCnfModGlobalsPermitted is used if there is limited access to modConfData_t variables.

I appreciate your help and feedback. Please provide further guidance in this.
Thanks!

rgerhards replied:
Thx for you approach - I'll happily look into it. But pls bear a little with me, I may take a day or two.

sourav-sharma796 replied:
Hi @rgerhards, did you check?

sourav-sharma796 commented on plugins/imrelp/imrelp.c:
```diff
@@ -183,10 +185,18 @@
-             "imrelp[%s]: error '%s', object "
-             " '%s' - input may not work as intended",
-             inst->pszBindPort, errmesg, objinfo);
+    if(!confFreed) {
+	    LogError(0, RS_RET_RELP_AUTH_FAIL, 
+	             "imrelp[%s]: error '%s', object "
+		     " '%s' - input may not work as intended",
+	             inst->pszBindPort, errmesg, objinfo);
+    }
+    else {
+	    LogError(0, RS_RET_RELP_AUTH_FAIL, 
+		     "imrelp[]: error '%s', object "
+	             " '%s' - input may not work as intended",
+	             errmesg, objinfo);
+    }
```
That's a good point. I'll optimize this to the suggested change.
Thanks!

rgerhards commented:
sry, I am swamped. It try this week, else it has to wait till mid august. But I am sure we can merge in the 8.2508 milestone. It's almost done - thx again!

rgerhards commented:
sry, I need to move it a bit, but it'll hit the daily stable build pretty soon. Too much was going on :-( My apologies

rgerhards commented:
@codex review

sourav-sharma796 commented:
@rgerhards,

I have made all the changes, and I think that now this PR is ready to be merged. I welcome you to make any change if required during merge.

Thank you for your support and help. It's great working with you.

rgerhards commented:
I'll now merge as a pragmatic decision. Will look deeper into it. Once again thanks for bringing this up. Shall I ping you if I come up with a different solution (I think there is)?

rgerhards commented:
oh, we need one run of 
```
$ devtools/format-code.sh
```
To make the style matching. I overlooked that check went red.

sourav-sharma796 commented:
> I'll now merge as a pragmatic decision. Will look deeper into it. Once again thanks for bringing this up. Shall I ping you if I come up with a different solution (I think there is)?

Yes, please I'm open for further discussions and improvements on this. 
Thank you!

sourav-sharma796 commented:
> during the analysis, I may have found the actual root cause for the segfault. I would appreciate if you could have a look at #6053 - maybe you can even check if it works in your environment.
> 
> If you would like to take it from my branch, that's fine with me - after all, you did by far the most work on identifying and solving this issue. I just added a bit of internal knowledge. But I am not 100% sure yet if my one is fully correct.

Hi, @rgerhards I'll test it out this fix in my setup environment and inform you if it works. Btw, I've fixes code style errors by running the repo tool.

Thanks!

rgerhards commented:
@sourav-sharma796 Thanks again for all your work. If you can verify the root cause is fixed, you should really copy over my small patch - the credits definitely are deserved by you!

rgerhards reviewed (approved):
pragmatic "ok to merge", best after alternative approach has also been tried.

sourav-sharma796 commented:
Hi @rgerhards,

I've tested the fix provided at https://github.com/rsyslog/rsyslog/pull/6053 and it works fine.

I had tried moving rsconf destruct in deinitAll after module unload which didn't work due to some issue but didn't think about just moving relpEngineDestruct call.

Thanks for your precise solution.

would you like me to rebase this branch with the proposed fix?

sourav-sharma796 commented:
Hi @rgerhards, I've rebased this branch and dropped other previous commits. New commit  https://github.com/rsyslog/rsyslog/pull/5600/commits/cb7a4630f29e7dd37e3846360fb9e6e9ce7b23b4 has been added.

Please ignore the PR closed thing, it automatically closed when I synced branch with main.

Please let me know any other task if required to complete this PR merge.

Thanks!
